RecoLocalCalo/HGCalRecProducers/plugins/HeterogeneousHGCalHEFCellPositionsConditions.cc

```diff
@@ -0,0 +1,301 @@
+#include "CondFormats/HGCalObjects/interface/HeterogeneousHGCalHEFCellPositionsConditions.h"
+
+HeterogeneousHGCalHEFCellPositionsConditions::HeterogeneousHGCalHEFCellPositionsConditions(
+    cpos::HGCalPositionsMapping* cpuPos) {
+  //HGCalPositions as defined in hgcal_conditions::positions
+  this->sizes_ = calculate_memory_bytes_(cpuPos);
+  this->chunk_ = allocate_memory_(this->sizes_);
+  transfer_data_to_heterogeneous_pointers_(this->sizes_, cpuPos);
+  transfer_data_to_heterogeneous_vars_(cpuPos);
+}
+
+size_t HeterogeneousHGCalHEFCellPositionsConditions::allocate_memory_(const std::vector<size_t>& sz) {
+  size_t chunk = std::accumulate(sz.begin(), sz.end(), 0);  //total memory required in bytes
+  cudaCheck(cudaMallocHost(&this->posmap_.x, chunk));
+  return chunk;
+}
+
+void HeterogeneousHGCalHEFCellPositionsConditions::transfer_data_to_heterogeneous_pointers_(
+    const std::vector<size_t>& sz, cpos::HGCalPositionsMapping* cpuPos) {
+  //store cumulative sum in bytes and convert it to sizes in units of C++ typesHEF, i.e., number if items to be transferred to GPU
+  std::vector<size_t> cumsum_sizes(sz.size() + 1, 0);  //starting with zero
+  std::partial_sum(sz.begin(), sz.end(), cumsum_sizes.begin() + 1);
+  for (unsigned int i = 1; i < cumsum_sizes.size(); ++i)  //start at second element (the first is zero)
+  {
+    size_t types_size = 0;
+    if (cpos::types[i - 1] == cpos::HeterogeneousHGCalPositionsType::Float)
+      types_size = sizeof(float);
+    else if (cpos::types[i - 1] == cpos::HeterogeneousHGCalPositionsType::Int32_t)
+      types_size = sizeof(int32_t);
+    else if (cpos::types[i - 1] == cpos::HeterogeneousHGCalPositionsType::Uint32_t)
+      types_size = sizeof(uint32_t);
+    else
+      edm::LogError("HeterogeneousHGCalHEFCellPositionsConditions") << "Wrong HeterogeneousHGCalPositionsMapping type";
+    cumsum_sizes[i] /= types_size;
```

Comment:
This is not solved: still possible division by 0 here. Please fix (in other similar cases you throw)

Reply:
Fixed.